File: src/orchestrator/nl_engine.py
"""Natural language intent parsing (Korean + English) with cluster support."""
import re
from .models import IntentAction, ParsedIntent


# Patterns: (regex, action, group names for service/replicas/image/memory/cpu)
# Order matters: more specific patterns first
PATTERNS = [
    # Scale: "nginx를 5개로 스케일해줘", "scale nginx to 5"
    (r"(?:스케일|스케일링|늘려|줄여).*?(\w+).*?(\d+)\s*개", IntentAction.SCALE, ("service", "replicas")),
    (r"(\w+)\s*(?:를|을)\s*(\d+)\s*개(?:로)?\s*(?:스케일|실행)", IntentAction.SCALE, ("service", "replicas")),
    (r"(?:scale|replicate)\s+(\w+)\s+(?:to|up to)\s*(\d+)", IntentAction.SCALE, ("service", "replicas")),
    (r"(\w+)\s*(\d+)\s*개\s*(?:로)?\s*스케일", IntentAction.SCALE, ("service", "replicas")),
    # Migrate: "nginx를 node-b로 마이그레이션해줘", "migrate nginx to node-b"
    (r"(\w[\w-]*)\s*(?:를|을)\s*(\w[\w-]*)\s*(?:로|으로)\s*(?:마이그레이션|이동|옮겨|이전)", IntentAction.MIGRATE, ("service", "target_node")),
    (r"(?:마이그레이션|이동|옮겨)\s+(\w[\w-]*)\s+(\w[\w-]*)", IntentAction.MIGRATE, ("service", "target_node")),
    (r"(?:migrate|move)\s+(\w[\w-]*)\s+(?:to|from\s+\w+\s+to)\s+(\w[\w-]*)", IntentAction.MIGRATE, ("service", "target_node")),
    # Drain: "node-a 드레인해줘", "drain node-a"
    (r"(\w[\w-]*)\s*(?:를|을)?\s*(?:드레인|비워|배출)", IntentAction.DRAIN, ("target_node", None)),
    (r"(?:drain|evacuate)\s+(\w[\w-]*)", IntentAction.DRAIN, ("target_node", None)),
    # Deploy: "webapp 배포 이미지 myapp:v1" before "X 배포", then "nginx 배포해줘", "deploy nginx"
    (r"(\w+)\s*배포.*?이미지\s+(\S+)", IntentAction.DEPLOY, ("name", "image")),
    (r"(\w+)\s*(?:서비스\s*)?(?:를\s*)?배포(?:\s*해줘)?", IntentAction.DEPLOY, ("name", None)),
    (r"deploy\s+(\w+)(?:\s+(\S+))?", IntentAction.DEPLOY, ("name", "image")),
    # Resource: "redis 메모리 512MB 제한", "nginx에 메모리 256"
    (r"([a-zA-Z0-9_-]+).*?메모리\s*(\d+)\s*(m|mb|g|gb)?", IntentAction.RESOURCE, ("service", "memory")),
    (r"([a-zA-Z0-9_-]+)\s*에\s*메모리\s*(\d+)\s*(m|mb|g|gb)?", IntentAction.RESOURCE, ("service", "memory")),
    (r"(?:memory|mem)\s+(?:of\s+)?(\w+)\s+(?:to\s+)?(\d+)\s*(m|mb|g|gb)?", IntentAction.RESOURCE, ("service", "memory")),
    (r"(\w+).*?cpu\s*(\d*\.?\d+)", IntentAction.RESOURCE, ("service", "cpu")),
    # Stop: "nginx 컨테이너 종료해줘", "nginx 중지", "stop redis"
    (r"([a-zA-Z0-9_-]+)\s*컨테이너\s*(?:종료|중지)(?:\s*해줘)?", IntentAction.STOP, ("service", None)),
    (r"(\w+)\s*(?:를|을)?\s*중지", IntentAction.STOP, ("service", None)),
    (r"stop\s+(\w+)", IntentAction.STOP, ("service", None)),
    # Cluster status
    (r"클러스터\s*(?:상태|상황|현황|정보)|cluster\s*(?:status|info)", IntentAction.CLUSTER_STATUS, ()),
    # Node list
    (r"노드\s*(?:목록|리스트|상태)|node[s]?\s*(?:list|status)", IntentAction.NODE_LIST, ()),
    # List
    (r"목록|리스트|서비스\s*목록|list\s*services?", IntentAction.LIST, ()),
]

# Normalize memory suffix to docker format (e.g. 512m, 1g)
def _norm_memory(num: str, unit: str) -> str:
    u = (unit or "m").lower().replace("mb", "m").replace("gb", "g")
    if u not in ("m", "g"):
        u = "m"
    return f"{num}{u}"


def parse(command: str) -> ParsedIntent:
    """Parse natural language command into ParsedIntent."""
    raw = command.strip()
    if not raw:
        return ParsedIntent(action=IntentAction.UNKNOWN, raw=raw)

    lower = raw.lower()
    for pattern, action, groups in PATTERNS:
        m = re.search(pattern, raw, re.IGNORECASE)
        if not m:
            continue
        gs = m.groups()
        intent = ParsedIntent(action=action, raw=raw)

        if "service" in groups:
            idx = groups.index("service")
            if idx < len(gs) and gs[idx]:
                intent.service_name = gs[idx].strip()
        if "replicas" in groups:
            idx = groups.index("replicas")
            if idx < len(gs) and gs[idx]:
                try:
                    intent.replicas = int(gs[idx])
                except ValueError:
                    pass
        if "name" in groups:
            idx = groups.index("name")
            if idx < len(gs) and gs[idx]:
                intent.service_name = gs[idx].strip()
        if "image" in groups:
            idx = groups.index("image")
            if idx < len(gs) and gs[idx]:
                intent.image = gs[idx].strip()
        if "memory" in groups and action == IntentAction.RESOURCE:
            if len(gs) >= 2 and gs[1]:
                intent.memory = _norm_memory(gs[1], gs[2] if len(gs) > 2 else "m")
        if "cpu" in groups:
            idx = groups.index("cpu")
            if idx < len(gs) and gs[idx]:
                intent.cpu = gs[idx].strip()
        if "target_node" in groups:
            idx = groups.index("target_node")
            if idx < len(gs) and gs[idx]:
                intent.target_node = gs[idx].strip()

        return intent

    # Fallback: single word might be "list" or service name for deploy
    words = raw.split()
    if len(words) == 1 and words[0].lower() in ("list", "목록", "리스트"):
        return ParsedIntent(action=IntentAction.LIST, raw=raw)
    if len(words) == 1 and words[0].lower() not in ("list", "목록"):
        return ParsedIntent(
            action=IntentAction.DEPLOY,
            service_name=words[0],
            image=None,
            raw=raw,
        )

    return ParsedIntent(action=IntentAction.UNKNOWN, raw=raw)

File: internal/nlengine/nlengine.go
package nlengine

import (
	"regexp"
	"strconv"
	"strings"

	"ai-container-go/internal/models"
)

type patternEntry struct {
	re     *regexp.Regexp
	action models.IntentAction
	groups []string
}

var patterns = []patternEntry{
	// Scale
	{regexp.MustCompile(`(?i)(?:스케일|스케일링|늘려|줄여).*?(\w+).*?(\d+)\s*개`), models.IntentScale, []string{"service", "replicas"}},
	{regexp.MustCompile(`(?i)(\w+)\s*(?:를|을)\s*(\d+)\s*개(?:로)?\s*(?:스케일|실행)`), models.IntentScale, []string{"service", "replicas"}},
	{regexp.MustCompile(`(?i)(?:scale|replicate)\s+(\w+)\s+(?:to|up to)\s*(\d+)`), models.IntentScale, []string{"service", "replicas"}},
	{regexp.MustCompile(`(?i)(\w+)\s*(\d+)\s*개\s*(?:로)?\s*스케일`), models.IntentScale, []string{"service", "replicas"}},
	// Migrate
	{regexp.MustCompile(`(?i)(\w[\w-]*)\s*(?:를|을)\s*(\w[\w-]*)\s*(?:로|으로)\s*(?:마이그레이션|이동|옮겨|이전)`), models.IntentMigrate, []string{"service", "target_node"}},
	{regexp.MustCompile(`(?i)(?:마이그레이션|이동|옮겨)\s+(\w[\w-]*)\s+(\w[\w-]*)`), models.IntentMigrate, []string{"service", "target_node"}},
	{regexp.MustCompile(`(?i)(?:migrate|move)\s+(\w[\w-]*)\s+(?:to|from\s+\w+\s+to)\s+(\w[\w-]*)`), models.IntentMigrate, []string{"service", "target_node"}},
	// Drain
	{regexp.MustCompile(`(?i)(\w[\w-]*)\s*(?:를|을)?\s*(?:드레인|비워|배출)`), models.IntentDrain, []string{"target_node"}},
	{regexp.MustCompile(`(?i)(?:drain|evacuate)\s+(\w[\w-]*)`), models.IntentDrain, []string{"target_node"}},
	// Deploy — with replicas + optional spread/node
	{regexp.MustCompile(`(?i)(\w+)(?:를|을)?\s*(?:분산\s*(?:해서\s*)?)?(\d+)\s*개\s*배포(?:\s*해줘)?`), models.IntentDeploy, []string{"name", "replicas"}},
	{regexp.MustCompile(`(?i)(\w+)(?:를|을)?\s*(\w[\w-]*)\s*(?:에|노드에)\s*(?:분산\s*)?(\d+)\s*개\s*배포(?:\s*해줘)?`), models.IntentDeploy, []string{"name", "target_node", "replicas"}},
	{regexp.MustCompile(`(?i)(\w+)(?:를|을)?\s*(\w[\w-]*)\s*(?:에|노드에)\s*배포(?:\s*해줘)?`), models.IntentDeploy, []string{"name", "target_node"}},
	{regexp.MustCompile(`(?i)(\w+)\s*(?:분산\s*)?배포.*?이미지\s+(\S+)`), models.IntentDeploy, []string{"name", "image"}},
	{regexp.MustCompile(`(?i)(\w+)\s*(?:서비스\s*)?(?:를\s*)?(?:분산\s*)?배포(?:\s*해줘)?`), models.IntentDeploy, []string{"name"}},
	{regexp.MustCompile(`(?i)deploy\s+(\w+)(?:\s+(\d+))?(?:\s+(\S+))?`), models.IntentDeploy, []string{"name", "replicas", "image"}},
	// Resource
	{regexp.MustCompile(`(?i)([a-zA-Z0-9_-]+).*?메모리\s*(\d+)\s*(m|mb|g|gb)?`), models.IntentResource, []string{"service", "memory"}},
	{regexp.MustCompile(`(?i)([a-zA-Z0-9_-]+)\s*에\s*메모리\s*(\d+)\s*(m|mb|g|gb)?`), models.IntentResource, []string{"service", "memory"}},
	{regexp.MustCompile(`(?i)(?:memory|mem)\s+(?:of\s+)?(\w+)\s+(?:to\s+)?(\d+)\s*(m|mb|g|gb)?`), models.IntentResource, []string{"service", "memory"}},
	{regexp.MustCompile(`(?i)(\w+).*?cpu\s*(\d*\.?\d+)`), models.IntentResource, []string{"service", "cpu"}},
	// Stop
	{regexp.MustCompile(`(?i)([a-zA-Z0-9_-]+)\s*컨테이너\s*(?:종료|중지)(?:\s*해줘)?`), models.IntentStop, []string{"service"}},
	{regexp.MustCompile(`(?i)(\w+)\s*(?:를|을)?\s*중지`), models.IntentStop, []string{"service"}},
	{regexp.MustCompile(`(?i)stop\s+(\w+)`), models.IntentStop, []string{"service"}},
	// Cluster status
	{regexp.MustCompile(`(?i)클러스터\s*(?:상태|상황|현황|정보)|cluster\s*(?:status|info)`), models.IntentClusterStatus, []string{}},
	// Node list
	{regexp.MustCompile(`(?i)노드\s*(?:목록|리스트|상태)|node[s]?\s*(?:list|status)`), models.IntentNodeList, []string{}},
	// List
	{regexp.MustCompile(`(?i)목록|리스트|서비스\s*목록|list\s*services?`), models.IntentList, []string{}},
}

// normMemory normalizes memory value and unit into a short form.
// e.g. "512" + "mb" -> "512m", "1" + "gb" -> "1g", "256" + "" -> "256m"
func normMemory(value, unit string) string {
	unit = strings.ToLower(strings.TrimSpace(unit))
	switch unit {
	case "g", "gb":
		return value + "g"
	case "m", "mb":
		return value + "m"
	default:
		return value + "m"
	}
}

// Parse parses a natural language command (Korean or English) and returns
// a ParsedIntent describing the detected action and parameters.
func Parse(command string) models.ParsedIntent {
	intent := models.ParsedIntent{
		Action: models.IntentUnknown,
		Raw:    command,
	}

	trimmed := strings.TrimSpace(command)

	for _, p := range patterns {
		match := p.re.FindStringSubmatch(trimmed)
		if match == nil {
			continue
		}

		intent.Action = p.action

		for i, groupName := range p.groups {
			captureIdx := i + 1
			if captureIdx >= len(match) {
				continue
			}
			val := match[captureIdx]
			if val == "" {
				continue
			}

			switch groupName {
			case "service":
				intent.ServiceName = val
			case "replicas":
				if n, err := strconv.Atoi(val); err == nil {
					intent.Replicas = &n
				}
			case "name":
				intent.ServiceName = val
			case "image":
				intent.Image = val
			case "memory":
				// For resource action, the memory value is this group
				// and the unit is the next capture group
				unit := ""
				if captureIdx+1 < len(match) {
					unit = match[captureIdx+1]
				}
				intent.Memory = normMemory(val, unit)
			case "cpu":
				intent.CPU = val
			case "target_node":
				intent.TargetNode = val
			}
		}

		return intent
	}

	// Fallback: single-word handling
	lower := strings.ToLower(trimmed)
	if lower == "list" || lower == "목록" || lower == "리스트" {
		intent.Action = models.IntentList
		return intent
	}

	// Single word -> treat as deploy with that word as service name
	if len(strings.Fields(trimmed)) == 1 && trimmed != "" {
		intent.Action = models.IntentDeploy
		intent.ServiceName = trimmed
		return intent
	}

	return intent
}
